README.md
[![Build Status](https://travis-ci.org/AASHE/django-constant-contact.svg?branch=master)](https://travis-ci.org/AASHE/django-constant-contact)
[![Coverage Status](https://coveralls.io/repos/AASHE/django-constant-contact/badge.svg?branch=master)](https://coveralls.io/r/AASHE/django-constant-contact?branch=master)

# django-constant-contact

Django package for creating email marketing campaigns in Constant Contact

## Installation

Using pip:

    pip install django-constant-contact

## Settings

The following variables should be set in your settings.py:

    CONSTANT_CONTACT_API_KEY
    CONSTANT_CONTACT_ACCESS_TOKEN
    CONSTANT_CONTACT_FROM_EMAIL
    CONSTANT_CONTACT_REPLY_TO_EMAIL
    CONSTANT_CONTACT_USERNAME
    CONSTANT_CONTACT_PASSWORD

`CONSTANT_CONTACT_API_KEY` is assigned by Constant Contact when
you register your application. (You need to register as a
Constant Contact developer.)

`CONSTANT_CONTACT_ACCESS_TOKEN` is the access token granted to
your application by a Constant Contact user. This is the User
who will own all Constant Contact versions of uploaded Issues.

## Usage Examples

Create a new marketing campaign:

    email_content = "<html>....</html>"
    text_content = "blah blah blah"
    constant_contact = ConstantContact()

    options = {name: , from_email: , email_content: , text_content:, ....}

    try:
      campaign = constant_contact.new_email_marketing_campaign(options)
    except ConstantContactAPIError as exc:
      print exc.errors

Update an existing campaign (stored in a model):

    options['email_marketing_campaign'] = campaign
    options['name'] = "new name"

    try:
      constant_contact.update_email_marketing_campaign(options)
    except ConstantContactAPIError as exc:
      print exc.errors


CHANGELOG.md
# Change Log

## [Unreleased]

### Changed
## [1.3] - 2017-01-07
- Update version of Django.

### Added
## [1.2] - 2016-09-07
- Implement work-around for Constant Contact known issue.  When
  triggered, it causes the Advanced Editor to display nothing.

### Changed
## [1.1.3] - 2016-08-18
- Upgrade version of premailer used.

### Changed
## [1.1.2] - 2016-08-09
- Upgrade pip and setuptools in .travis.yml. django-bulletin
  depends on django-minify, which depends on other
  packages which need the upgraded setuptools.

### Added
## [1.1] - 2016-08-09
- Email content is now html-minified before sending to Constant Contact.

### Changed
## [1.0.5] - 2016-03-18
- Don't allow Django 1.9rcx to be installed.

### Changed
## [1.0.4] - 2016-03-18
- Set zip_safe = False in setup.py.

### Changed
## [1.0.2] - 2016-03-18
- Replaced README.md with README.rst.

### Changed
## [1.0.1] - 2016-03-18
- Removed Mailchimp dependency.


run_tests.py
#!/usr/bin/env python
import os
import sys
import django

BASE_PATH = os.path.dirname(__file__)


def main():
    """
    Standalone django model test with a 'memory-only-django-installation'.
    You can play with a django model without a complete django app installation
    http://www.djangosnippets.org/snippets/1044/
    """
    sys.exc_clear()

    os.environ["DJANGO_SETTINGS_MODULE"] = "django.conf.global_settings"
    from django.conf import global_settings

    global_settings.INSTALLED_APPS = (
        'django.contrib.auth',
        'django.contrib.sessions',
        'django.contrib.contenttypes',
        'django_constant_contact',
    )
    if django.VERSION > (1, 2):
        global_settings.DATABASES = {
            'default': {
                'ENGINE': 'django.db.backends.sqlite3',
                'NAME': os.path.join(BASE_PATH, 'connpass.sqlite'),
                'USER': '',
                'PASSWORD': '',
                'HOST': '',
                'PORT': '',
            }
        }
    else:
        global_settings.DATABASE_ENGINE = "sqlite3"
        global_settings.DATABASE_NAME = ":memory:"

    global_settings.MIDDLEWARE_CLASSES = (
        'django.middleware.common.CommonMiddleware',
        'django.contrib.sessions.middleware.SessionMiddleware',
        'django.middleware.csrf.CsrfViewMiddleware',
        'django.contrib.auth.middleware.AuthenticationMiddleware',
        'django.contrib.messages.middleware.MessageMiddleware'
    )

    # custom settings for tests
    global_settings.CONSTANT_CONTACT_API_KEY = os.environ.get(
        'CONSTANT_CONTACT_API_KEY', None)
    global_settings.CONSTANT_CONTACT_ACCESS_TOKEN = os.environ.get(
        'CONSTANT_CONTACT_ACCESS_TOKEN', None)
    global_settings.CONSTANT_CONTACT_FROM_EMAIL = os.environ.get(
        'CONSTANT_CONTACT_FROM_EMAIL', None)
    global_settings.CONSTANT_CONTACT_REPLY_TO_EMAIL = os.environ.get(
        'CONSTANT_CONTACT_REPLY_TO_EMAIL', None)
    global_settings.CONSTANT_CONTACT_USERNAME = os.environ.get(
        'CONSTANT_CONTACT_USERNAME', None)
    global_settings.CONSTANT_CONTACT_PASSWORD = os.environ.get(
        'CONSTANT_CONTACT_PASSWORD', None)

    global_settings.SECRET_KEY = "blahblah"

    from django.test.utils import get_runner
    test_runner = get_runner(global_settings)

    django.setup()

    if django.VERSION > (1, 2):
        test_runner = test_runner()
        failures = test_runner.run_tests(['django_constant_contact', ])
    else:
        failures = test_runner(['django_constant_contact', ], verbosity=1)
    sys.exit(failures)

if __name__ == '__main__':
    main()


django_constant_contact/tests.py
import time
import uuid
import unittest

from django.conf import settings
import django.test

from .models import (ConstantContact,
                     ConstantContactAPIError,
                     EmailMarketingCampaign)


ORG_ADDRESS = {
    'organization_name': 'My Organization',
    'address_line_1': '123 Maple Street',
    'address_line_2': 'Suite 1',
    'address_line_3': '',
    'city': 'Anytown',
    'state': 'MA',
    'international_state': '',
    'postal_code': '01444',
    'country': 'US'
}


# Define assertIn here, since it doesn't live on unittest.TestCase
# until Python 2.7.
def assertIn(self, test_value, expected_set):
    msg = "%s did not occur in %s" % (test_value, expected_set)
    self.assert_(test_value in expected_set, msg)


class ConstantContactTests(unittest.TestCase):

    def setUp(self):
        self.cc = ConstantContact()
        self.email_marketing_campaign = None
        self.email_marketing_campaign_kwargs = {
            'name': 'Test Campaign {0}'.format(uuid.uuid4()),
            'email_content': '<html><body>Test Email Content</body></html>',
            'from_email': settings.CONSTANT_CONTACT_FROM_EMAIL,
            'from_name': 'Test Sender',
            'reply_to_email': settings.CONSTANT_CONTACT_REPLY_TO_EMAIL,
            'subject': 'Test Subject',
            'text_content': '<text>Test Text Content</text>',
            'address': ORG_ADDRESS}
        time.sleep(1)  # So we don't run over our queries-per-second quota.

    def tearDown(self):
        if self.email_marketing_campaign:
            # Call delete on any email marketing campaign
            # we created, which should also delete it up
            # on Constant Contact's servers.
            self.email_marketing_campaign.delete()

    def test_connect(self):
        response = self.cc.api.get('/account/info')
        self.assertEqual(200, response.status_code)

    def test_create_email_marketing_campaign(self):
        self.email_marketing_campaign = self.cc.new_email_marketing_campaign(
            **self.email_marketing_campaign_kwargs)
        assert isinstance(self.email_marketing_campaign,
                          EmailMarketingCampaign)

    def test_update_email_marketing_campaign(self):
        """Can we update an Email Marketing Campaign?
        Will fail if test_create_email_marketing_campaign fails.
        """
        self.email_marketing_campaign = self.cc.new_email_marketing_campaign(
            **self.email_marketing_campaign_kwargs)

        update_kwargs = self.email_marketing_campaign_kwargs
        updated_subject = (
            self.email_marketing_campaign.data['subject'] + 'Tom Brady')
        update_kwargs['subject'] = updated_subject
        update_kwargs['email_marketing_campaign'] = (
            self.email_marketing_campaign)

        self.email_marketing_campaign = (
            self.cc.update_email_marketing_campaign(**update_kwargs))

        self.assertEqual(updated_subject,
                         self.email_marketing_campaign.data['subject'])

    def test_delete_email_marketing_campaign(self):
        """Can we delete an email marketing campaign?
        """
        email_marketing_campaign = self.cc.new_email_marketing_campaign(
            **self.email_marketing_campaign_kwargs)
        self.cc.delete_email_marketing_campaign(
            email_marketing_campaign)

    def test_failed_delete_email_marketing_campaign_raises_exception(self):
        """When a delete fails, is an exception raised?
        """
        # Try to delete a bogus ID:
        email_marketing_campaign = EmailMarketingCampaign.objects.create(
            data={'id': 1})
        self.assertRaises(ConstantContactAPIError,
                          self.cc.delete_email_marketing_campaign,
                          email_marketing_campaign)

    def test_server_version_is_removed_upon_delete_of_email_marketing_campaign(
            self):
        """When EmailMarketingCampaign deleted, is the CC version deleted?
        """
        email_marketing_campaign = self.cc.new_email_marketing_campaign(
            **self.email_marketing_campaign_kwargs)
        # GET the email marketing campaign:
        cc = ConstantContact()
        url = cc.api.join('/'.join([
            cc.EMAIL_MARKETING_CAMPAIGN_URL,
            str(email_marketing_campaign.data['id'])]))
        response = url.get()
        self.assertEqual(200, response.status_code)
        # Delete the EmailMarketingCampaign:
        email_marketing_campaign.delete()
        # Same GET should 404 now:
        response = url.get()
        self.assertEqual(404, response.status_code)

    def test_inline_css(self):
        """Can we inline CSS?
        """
        html = """
        <html>
          <head>
            <style type="text/css">
             body {
               width: 100%;
             }
             a {
               text-decoration: none;
             }
             .section {
               padding-top: 5px;
             }
            </style>
          </head>
          <body>
            <div class="section">
              <a href="#"></a>
            </div>
          </body>
        </html>
        """
        expected_inlined_html = """
        <html>
          <head>
          </head>
          <body style="width: 100%">
            <div class="section" style="padding-top: 5px">
              <a href="#" style="text-decoration: none"></a>
            </div>
          </body>
        </html>
        """
        inlined_html = self.cc.inline_css(html)
        whitespaceless_expected_inlined_html = (
            ''.join(expected_inlined_html.split(expected_inlined_html)))
        whitespaceless_inlined_html = ''.join(inlined_html.split(inlined_html))

        self.assertEqual(whitespaceless_expected_inlined_html,
                         whitespaceless_inlined_html)

    def test_preview_email_marketing_campaign(self):
        """Can we preview an Email Marketing Campaign?

        Pretty weak test.  Makes no assertions.
        """
        self.email_marketing_campaign = self.cc.new_email_marketing_campaign(
            **self.email_marketing_campaign_kwargs)
        html, text = self.cc.preview_email_marketing_campaign(
            self.email_marketing_campaign)

    def test_create_view_as_webpage(self):
        """Can we set the "view as webpage" flag when we create an EMC?
        """
        link_text = 'bzxbnxzcvxczvnbmxcvzxcvcvxnbv'
        page_text = 'uioqweriuoerwiueriuerwuiweriu'

        kwargs = self.email_marketing_campaign_kwargs
        kwargs['is_view_as_webpage_enabled'] = True
        kwargs['view_as_web_page_link_text'] = link_text
        kwargs['view_as_web_page_text'] = page_text
        self.email_marketing_campaign = self.cc.new_email_marketing_campaign(
            **kwargs)
        html, text = self.cc.preview_email_marketing_campaign(
            self.email_marketing_campaign)
        assertIn(self, link_text, html)
        assertIn(self, page_text, html)

    def test_update_view_as_webpage(self):
        """Can we set the "view as webpage" flag when we update an EMC?
        """
        self.email_marketing_campaign = self.cc.new_email_marketing_campaign(
            **self.email_marketing_campaign_kwargs)

        link_text = 'bzxbnxzcvxczvnbmxcvzxcvcvxnbv'
        page_text = 'uioqweriuoerwiueriuerwuiweriu'

        kwargs = self.email_marketing_campaign_kwargs
        kwargs['email_marketing_campaign'] = self.email_marketing_campaign
        kwargs['is_view_as_webpage_enabled'] = True
        kwargs['view_as_web_page_link_text'] = link_text
        kwargs['view_as_web_page_text'] = page_text
        self.email_marketing_campaign = (
            self.cc.update_email_marketing_campaign(**kwargs))
        html, text = self.cc.preview_email_marketing_campaign(
            self.email_marketing_campaign)
        assertIn(self, link_text, html)
        assertIn(self, page_text, html)

    def test_create_set_permission_reminder(self):
        """Can we set the permission reminder flag when we create an EMC?
        """
        reminder_text = '45646564788794561232456786453'

        kwargs = self.email_marketing_campaign_kwargs
        kwargs['is_permission_reminder_enabled'] = True
        kwargs['permission_reminder_text'] = reminder_text
        self.email_marketing_campaign = self.cc.new_email_marketing_campaign(
            **kwargs)
        html, text = self.cc.preview_email_marketing_campaign(
            self.email_marketing_campaign)
        assertIn(self, reminder_text, html)

    def test_update_set_permission_reminder(self):
        """Can we set the permission reminder flag when we update an EMC?
        """
        self.email_marketing_campaign = self.cc.new_email_marketing_campaign(
            **self.email_marketing_campaign_kwargs)

        reminder_text = '45646564788794561232456786453'

        kwargs = self.email_marketing_campaign_kwargs
        kwargs['email_marketing_campaign'] = self.email_marketing_campaign
        kwargs['is_permission_reminder_enabled'] = True
        kwargs['permission_reminder_text'] = reminder_text
        self.email_marketing_campaign = (
            self.cc.update_email_marketing_campaign(**kwargs))
        html, text = self.cc.preview_email_marketing_campaign(
            self.email_marketing_campaign)
        assertIn(self, reminder_text, html)


class EmailMarketingCampaignTests(django.test.TestCase):

    def test_pre_save_works(self):
        """Does pre_save set constant_contact_id?
        """
        data = {'id': 1}
        emc = EmailMarketingCampaign(data=data)
        emc.save()
        self.assertEqual(emc.constant_contact_id, '1')

    def test_pre_delete_fires(self):
        """Is pre_delete wired up correctly?

        If yes, it should raise an exception when we try to delete
        an EmailMarketingCampaign with a bogus constant_contact_id.
        """
        data = {'id': 1}
        emc = EmailMarketingCampaign(data=data)
        emc.save()

        self.assertRaises(ConstantContactAPIError, emc.delete)


django_constant_contact/utils.py
import mailchimp


CONSTANT_CONTACT_COUNTRIES = [
    'AF', 'AX', 'AL', 'DZ', 'AS', 'AD', 'AO', 'AI', 'AQ', 'AG', 'AR', 'AM',
    'AW', 'AU', 'AT', 'AZ', 'BS', 'BH', 'BD', 'BB', 'BY', 'BE', 'BZ', 'BJ',
    'BM', 'BT', 'BO', 'BA', 'BW', 'BV', 'BR', 'IO', 'BN', 'BG', 'BF', 'BI',
    'KH', 'CM', 'CA', 'CV', 'KY', 'CF', 'TD', 'CL', 'CN', 'CX', 'CC', 'CO',
    'KM', 'CG', 'CD', 'CK', 'CR', 'CI', 'HR', 'CY', 'CZ', 'DK', 'DJ', 'DM',
    'DO', 'TL', 'EC', 'EG', 'SV', 'U1', 'GQ', 'ER', 'EE', 'ET', 'FO', 'FK',
    'FJ', 'FI', 'FR', 'GF', 'PF', 'TF', 'GA', 'GM', 'GE', 'DE', 'GH', 'GI',
    'GR', 'GL', 'GD', 'GP', 'GU', 'GT', 'GG', 'GN', 'GW', 'GY', 'HT', 'HM',
    'HN', 'HK', 'HU', 'IS', 'IN', 'ID', 'IQ', 'IE', 'IM', 'IL', 'IT', 'JM',
    'JP', 'JE', 'JO', 'KZ', 'KE', 'KI', 'KW', 'KG', 'LA', 'LV', 'LB', 'LS',
    'LR', 'LY', 'LI', 'LT', 'LU', 'MO', 'MK', 'MG', 'MW', 'MY', 'MV', 'ML',
    'MT', 'MH', 'MQ', 'MR', 'MU', 'YT', 'MX', 'FM', 'MD', 'MC', 'MN', 'ME',
    'MS', 'MA', 'MZ', 'MM', 'NA', 'NR', 'NP', 'NL', 'AN', 'NT', 'NC', 'NZ',
    'NI', 'NE', 'NG', 'NU', 'NF', 'U4', 'MP', 'NO', 'OM', 'PK', 'PW', 'PS',
    'PA', 'PG', 'PY', 'PE', 'PH', 'PN', 'PL', 'PT', 'PR', 'QA', 'RE', 'RO',
    'RU', 'RW', 'BL', 'SH', 'KN', 'LC', 'MF', 'PM', 'VC', 'WS', 'SM', 'ST',
    'SA', 'U3', 'SN', 'RS', 'SC', 'SL', 'SG', 'SK', 'SI', 'SB', 'SO', 'ZA',
    'GS', 'KR', 'ES', 'LK', 'SR', 'SJ', 'SZ', 'SE', 'CH', 'TW', 'TJ', 'TZ',
    'TH', 'TG', 'TK', 'TO', 'TT', 'TN', 'TR', 'TM', 'TC', 'TV', 'UG', 'UA',
    'AE', 'GB', 'US', 'USA', 'UM', 'UY', 'UZ', 'VU', 'VA', 'VE', 'VN', 'VG',
    'VI', 'U2', 'WF', 'EH', 'YE', 'ZM', 'ZW']


class ConstantContactAPIError(Exception):
    """An exception that passes error info from response to exception catcher.
    """

    def __init__(self, response, *args, **kwargs):
        super(ConstantContactAPIError, self).__init__(*args, **kwargs)
        self.message = str(response.status_code) + ': ' + response.reason
        self.errors = json.loads(response.content)

    def __str__(self):
        s = '<ConstantContactAPIError ' + self.message + ' '
        for error in self.errors:
            s += str(error)
        s += '>'
        return s


class ConstantContact(object):

    API_URL = 'https://api.constantcontact.com/v2/'
    EMAIL_MARKETING_CAMPAIGN_URL = 'emailmarketing/campaigns'

    def __init__(self):
        self.api = nap.url.Url(
            self.API_URL,
            params={'api_key': settings.CONSTANT_CONTACT_API_KEY,
                    'access_token': settings.CONSTANT_CONTACT_ACCESS_TOKEN})

    def handle_response_status(self, response):
        """Raise an exception for response errors.

        Constant Contact returns some nice error info that we lose if
        we only response.raise_for_status(). So we trap most errors
        and raise exceptions that preserve the error info.

        Calling raise_for_status() catches any other errors.
        """
        if 400 <= response.status_code <= 599:
            raise ConstantContactAPIError(response)

        response.raise_for_status()

    def new_email_marketing_campaign(self, name, email_content, from_email,
                                     from_name, reply_to_email, subject,
                                     text_content, address):
        """Create a Constant Contact email marketing campaign.
        Returns an EmailMarketingCampaign object.
        """
        url = self.api.join(self.EMAIL_MARKETING_CAMPAIGN_URL)

        inlined_email_content = self.inline_css(email_content)

        data = {
            'name': name,
            'subject': subject,
            'from_name': from_name,
            'from_email': from_email,
            'reply_to_email': reply_to_email,
            'email_content': inlined_email_content,
            'email_content_format': 'HTML',
            'text_content': text_content,
            'message_footer': {
                'organization_name': address['organization_name'],
                'address_line_1': address['address_line_1'],
                'address_line_2': address['address_line_2'],
                'address_line_3': address['address_line_3'],
                'city': address['city'],
                'state': address['state'],
                'international_state': address['international_state'],
                'postal_code': address['postal_code'],
                'country': address['country']
            }
        }
        response = url.post(data=json.dumps(data),
                            headers={'content-type': 'application/json'})

        self.handle_response_status(response)

        return EmailMarketingCampaign.objects.create(data=response.json())

    def update_email_marketing_campaign(self, email_marketing_campaign,
                                        name, email_content, from_email,
                                        from_name, reply_to_email, subject,
                                        text_content, address):
        """Update a Constant Contact email marketing campaign.
        Returns the updated EmailMarketingCampaign object.
        """
        url = self.api.join(
            '/'.join([self.EMAIL_MARKETING_CAMPAIGN_URL,
                      str(email_marketing_campaign.constant_contact_id)]))

        inlined_email_content = self.inline_css(email_content)

        data = {
            'name': name,
            'subject': subject,
            'from_name': from_name,
            'from_email': from_email,
            'reply_to_email': reply_to_email,
            'email_content': inlined_email_content,
            'email_content_format': 'HTML',
            'text_content': text_content,
            'message_footer': {
                'organization_name': address['organization_name'],
                'address_line_1': address['address_line_1'],
                'address_line_2': address['address_line_2'],
                'address_line_3': address['address_line_3'],
                'city': address['city'],
                'state': address['state'],
                'international_state': address['international_state'],
                'postal_code': address['postal_code'],
                'country': address['country']
            }
        }
        response = url.put(data=json.dumps(data),
                           headers={'content-type': 'application/json'})

        self.handle_response_status(response)

        email_marketing_campaign.data = response.json()
        email_marketing_campaign.save()

        return email_marketing_campaign

    def delete_email_marketing_campaign(self, email_marketing_campaign):
        """Deletes a Constant Contact email marketing campaign.
        """
        url = self.api.join('/'.join([
            self.EMAIL_MARKETING_CAMPAIGN_URL,
            str(email_marketing_campaign.constant_contact_id)]))
        response = url.delete()
        self.handle_response_status(response)
        return response

    def inline_css(self, html):
        """Inlines CSS defined in external style sheets.

        This implementation requires a Mailchimp account.
        """
        mailchimp_api = mailchimp.Mailchimp(settings.MAILCHIMP_API_KEY)
        response = mailchimp_api.helper.inline_css(html)
        return response['html']

    def preview_email_marketing_campaign(self, email_marketing_campaign):
        """Returns HTML and text previews of an EmailMarketingCampaign.
        """
        url = self.api.join('/'.join([
            self.EMAIL_MARKETING_CAMPAIGN_URL,
            str(email_marketing_campaign.constant_contact_id),
            'preview']))
        response = url.get()
        self.handle_response_status(response)
        return (response.json()['preview_email_content'],
                response.json()['preview_text_content'])


django_constant_contact/models.py
# -*- coding: utf-8 -*-

import json

import jsonfield
import nap
from django.conf import settings
from django.db import models
from django.db.models.signals import pre_delete, pre_save
from htmlmin.minify import html_minify
from premailer import Premailer


class ConstantContactAPIError(Exception):
    """An exception that passes error info from response to exception catcher.
    """

    def __init__(self, response, *args, **kwargs):
        super(ConstantContactAPIError, self).__init__(*args, **kwargs)
        self.message = str(response.status_code) + ': ' + response.reason
        self.errors = json.loads(response.content)

    def __str__(self):
        s = '<ConstantContactAPIError ' + self.message + ' '
        for error in self.errors:
            s += str(error)
        s += '>'
        return s


class ConstantContact(object):

    API_URL = 'https://api.constantcontact.com/v2/'
    EMAIL_MARKETING_CAMPAIGN_URL = 'emailmarketing/campaigns'

    def __init__(self):
        self.api = nap.url.Url(
            self.API_URL,
            params={'api_key': settings.CONSTANT_CONTACT_API_KEY,
                    'access_token': settings.CONSTANT_CONTACT_ACCESS_TOKEN})

    def handle_response_status(self, response):
        """Raise an exception for response errors.

        Constant Contact returns some nice error info that we lose if
        we only response.raise_for_status(). So we trap most errors
        and raise exceptions that preserve the error info.

        Calling raise_for_status() catches any other errors.
        """
        if 400 <= response.status_code <= 599:
            raise ConstantContactAPIError(response)

        response.raise_for_status()

    def new_email_marketing_campaign(self, name, email_content, from_email,
                                     from_name, reply_to_email, subject,
                                     text_content, address,
                                     is_view_as_webpage_enabled=False,
                                     view_as_web_page_link_text='',
                                     view_as_web_page_text='',
                                     is_permission_reminder_enabled=False,
                                     permission_reminder_text=''):
        """Create a Constant Contact email marketing campaign.
        Returns an EmailMarketingCampaign object.
        """
        url = self.api.join(self.EMAIL_MARKETING_CAMPAIGN_URL)

        inlined_email_content = self.inline_css(email_content)
        minified_email_content = html_minify(inlined_email_content)
        worked_around_email_content = work_around(minified_email_content)

        data = {
            'name': name,
            'subject': subject,
            'from_name': from_name,
            'from_email': from_email,
            'reply_to_email': reply_to_email,
            'email_content': worked_around_email_content,
            'email_content_format': 'HTML',
            'text_content': text_content,
            'message_footer': {
                'organization_name': address['organization_name'],
                'address_line_1': address['address_line_1'],
                'address_line_2': address['address_line_2'],
                'address_line_3': address['address_line_3'],
                'city': address['city'],
                'state': address['state'],
                'international_state': address['international_state'],
                'postal_code': address['postal_code'],
                'country': address['country']
            },
            'is_view_as_webpage_enabled': is_view_as_webpage_enabled,
            'view_as_web_page_link_text': view_as_web_page_link_text,
            'view_as_web_page_text': view_as_web_page_text,
            'is_permission_reminder_enabled': is_permission_reminder_enabled,
            'permission_reminder_text': permission_reminder_text
        }

        response = url.post(data=json.dumps(data),
                            headers={'content-type': 'application/json'})

        self.handle_response_status(response)

        return EmailMarketingCampaign.objects.create(data=response.json())

    def update_email_marketing_campaign(self, email_marketing_campaign,
                                        name, email_content, from_email,
                                        from_name, reply_to_email, subject,
                                        text_content, address,
                                        is_view_as_webpage_enabled=False,
                                        view_as_web_page_link_text='',
                                        view_as_web_page_text='',
                                        is_permission_reminder_enabled=False,
                                        permission_reminder_text=''):
        """Update a Constant Contact email marketing campaign.
        Returns the updated EmailMarketingCampaign object.
        """
        url = self.api.join(
            '/'.join([self.EMAIL_MARKETING_CAMPAIGN_URL,
                      str(email_marketing_campaign.constant_contact_id)]))

        inlined_email_content = self.inline_css(email_content)
        minified_email_content = html_minify(inlined_email_content)
        worked_around_email_content = work_around(minified_email_content)

        data = {
            'name': name,
            'subject': subject,
            'from_name': from_name,
            'from_email': from_email,
            'reply_to_email': reply_to_email,
            'email_content': worked_around_email_content,
            'email_content_format': 'HTML',
            'text_content': text_content,
            'message_footer': {
                'organization_name': address['organization_name'],
                'address_line_1': address['address_line_1'],
                'address_line_2': address['address_line_2'],
                'address_line_3': address['address_line_3'],
                'city': address['city'],
                'state': address['state'],
                'international_state': address['international_state'],
                'postal_code': address['postal_code'],
                'country': address['country']
            },
            'is_view_as_webpage_enabled': is_view_as_webpage_enabled,
            'view_as_web_page_link_text': view_as_web_page_link_text,
            'view_as_web_page_text': view_as_web_page_text,
            'is_permission_reminder_enabled': is_permission_reminder_enabled,
            'permission_reminder_text': permission_reminder_text
        }

        response = url.put(data=json.dumps(data),
                           headers={'content-type': 'application/json'})

        self.handle_response_status(response)

        email_marketing_campaign.data = response.json()
        email_marketing_campaign.save()

        return email_marketing_campaign

    def delete_email_marketing_campaign(self, email_marketing_campaign):
        """Deletes a Constant Contact email marketing campaign.
        """
        url = self.api.join('/'.join([
            self.EMAIL_MARKETING_CAMPAIGN_URL,
            str(email_marketing_campaign.constant_contact_id)]))
        response = url.delete()
        self.handle_response_status(response)
        return response

    def inline_css(self, html):
        """Inlines CSS defined in external style sheets.
        """
        premailer = Premailer(html)
        inlined_html = premailer.transform(pretty_print=True)
        return inlined_html

    def preview_email_marketing_campaign(self, email_marketing_campaign):
        """Returns HTML and text previews of an EmailMarketingCampaign.
        """
        url = self.api.join('/'.join([
            self.EMAIL_MARKETING_CAMPAIGN_URL,
            str(email_marketing_campaign.constant_contact_id),
            'preview']))
        response = url.get()
        self.handle_response_status(response)
        return (response.json()['preview_email_content'],
                response.json()['preview_text_content'])


class EmailMarketingCampaign(models.Model):
    """A Constant Contact email marketing campaign.

    WARNING! Deleting an EmailMarketingCampaign will also delete
             the corresponding email marketing campaign at
             Constant Contact.

    The `data` field should be JSON from Constant Contact, describing
    the "real" email marketing campaign that lives on their server.
    (Constant Contact supplies this when an email marketing campaign
    is created or modified.)

    There should be a one-to-one relationship between each
    EmailMarketingCampaign and an email marketing campaign
    living on a Constant Contact server.  To enforce this, the
    ID Constant Contact uses to identify an email marketing campaign
    is stored here as constant_contact_id, with a uniqueness constraint.
    """
    constant_contact_id = models.BigIntegerField(unique=True)
    data = jsonfield.JSONField()

    @classmethod
    def pre_save(cls, sender, instance, *args, **kwargs):
        """Pull constant_contact_id out of data.
        """
        instance.constant_contact_id = str(instance.data['id'])

    @classmethod
    def pre_delete(cls, sender, instance, *args, **kwargs):
        """Deletes the CC email marketing campaign associated with me.
        """
        cc = ConstantContact()
        response = cc.delete_email_marketing_campaign(instance)
        response.raise_for_status()

pre_save.connect(EmailMarketingCampaign.pre_save,
                 sender=EmailMarketingCampaign)

pre_delete.connect(EmailMarketingCampaign.pre_delete,
                   sender=EmailMarketingCampaign)


def work_around(content):
    """A work-around for a Constant Contact known issue.

    ‎09-06-2016 02:28 PM, from Constant Contact support;

    >> It appears you are running into a known issue with creating
       emails with unsupported characters within them. In looking at
       the HTML that came over for this email, I found an instancea of
       a "smart" or "curly" apostrophe. Though it appears similar ( ’
       vs ' ), only the standard apostrophe (the second one) is
       available to be used.

    >> While this should return an error, the known issue is that we
       are creating the email with these special characters
       anyway. This issue has been brought to the attention of our
       engineering and development teams for investigation. The
       workaround at this point is remove any of these unsupported
       characters from the code before submitting.
    """
    return content.decode("utf-8").replace(u"\u2019", "'").encode("utf-8")
